packages/gatsby-plugin-image/src/codemod.js

```diff
@@ -0,0 +1,218 @@
+import graphql from "gatsby/graphql"
+import { parse, print } from "recast"
+import { transformFromAstSync, parseSync } from "@babel/core"
+
+const legacyFragments = [
+  `GatsbyImageSharpFixed`,
+  `GatsbyImageSharpFixed_withWebp`,
+  `GatsbyImageSharpFluid`,
+  `GatsbyImageSharpFluid_withWebp`,
+]
+
+const legacyFragmentsNoPlaceholder = [
+  `GatsbyImageSharpFixed_noBase64`,
+  `GatsbyImageSharpFixed_withWebp_noBase64`,
+  `GatsbyImageSharpFluid_noBase64`,
+  `GatsbyImageSharpFluid_withWebp_noBase64`,
+]
+
+const legacyFragmentsSVGPlaceholder = [
+  `GatsbyImageSharpFixed_tracedSVG`,
+  `GatsbyImageSharpFixed_withWebp_tracedSVG`,
+  `GatsbyImageSharpFluid_tracedSVG`,
+  `GatsbyImageSharpFluid_withWebp_tracedSVG`,
+]
+
+const typeMapper = {
+  fixed: `FIXED`,
+  fluid: `FLUID`,
+  constrained: `CONSTRAINED`,
+}
+
+export function babelRecast(code) {
+  const transformedAst = parse(code, {
+    parser: {
+      parse: source =>
+        parseSync(source, {
+          plugins: [
+            // [`@babel/plugin-syntax-typescript`, { isTSX: true }],
+            `@babel/plugin-syntax-jsx`,
+          ],
+        }),
+    },
+  })
+
+  const options = {
+    cloneInputAst: false,
+    code: false,
+    ast: true,
+    plugins: [updateImport],
+  }
+
+  const { ast } = transformFromAstSync(transformedAst, code, options)
+
+  const result = print(ast).code
+  return result
+}
+
+export default function updateImport({ types: t }) {
+  let imageImportName = `GatsbyImage`
+  return {
+    visitor: {
+      ImportDeclaration: path => {
+        const { node } = path
+        if (
+          node.source.value !== `gatsby-image` &&
+          node.source.value !== `gatsby-plugin-image/compat`
+        ) {
+          return
+        }
+        imageImportName = node.specifiers[0].local.name
+        const namedImport = t.importSpecifier(
+          t.identifier(`GatsbyImage`),
+          t.identifier(`GatsbyImage`)
+        )
+        const newImport = t.importDeclaration(
+          [namedImport],
+          t.stringLiteral(`gatsby-plugin-image`)
+        )
+        path.replaceWith(newImport)
+      },
+      JSXOpeningElement(path) {
+        const { node } = path
+        if (node.name.name !== imageImportName) {
+          return
+        }
+        const componentName = t.jsxIdentifier(`GatsbyImage`)
+
+        const otherAttributes = node.attributes.filter(
+          ({ name }) => name.name !== `fluid` && name.name !== `fixed`
+        )
+
+        // this expression is equivalent to data.file.childImageSharp.gatsbyImageData
+        const newImageExpression = t.memberExpression(
+          t.memberExpression(
+            t.memberExpression(t.identifier(`data`), t.identifier(`file`)),
+            t.identifier(`childImageSharp`)
+          ),
+          t.identifier(`gatsbyImageData`)
+        )
```

Comment:
I'm really nitpicking here, but instead of a comment to explain what it does this could be written as

```suggestion
          const newImageExpression = template.expression.ast`
            data.file.childImageSharp.gatsbyImageData
          `;
```

and you get `template` where you get `t`:
```js
export default function UpdateImport({ types: t, template }) {
```

Reply:
I noticed templates in some of the docs and hadn't dug into them yet. Definitely better!